runtime/autoload/provider/clipboard.vim

```diff
@@ -13,34 +14,79 @@ function! s:try_cmd(cmd, ...)
   return out
 endfunction
 
+let s:cache_enabled = 1
+
 if executable('pbcopy')
   let s:copy['+'] = 'pbcopy'
   let s:paste['+'] = 'pbpaste'
   let s:copy['*'] = s:copy['+']
   let s:paste['*'] = s:paste['+']
+  let s:cache_enabled = 0
 elseif executable('xclip')
-  let s:copy['+'] = 'xclip -i -selection clipboard'
+  let s:copy['+'] = 'xclip -quiet -i -selection clipboard'
   let s:paste['+'] = 'xclip -o -selection clipboard'
-  let s:copy['*'] = 'xclip -i -selection primary'
+  let s:copy['*'] = 'xclip -quiet -i -selection primary'
   let s:paste['*'] = 'xclip -o -selection primary'
 elseif executable('xsel')
-  let s:copy['+'] = 'xsel -i -b'
+  let s:copy['+'] = 'xsel --nodetach -i -b'
   let s:paste['+'] = 'xsel -o -b'
-  let s:copy['*'] = 'xsel -i -p'
+  let s:copy['*'] = 'xsel --nodetach -i -p'
   let s:paste['*'] = 'xsel -o -p'
 else
   echom 'clipboard: No shell command for communicating with the clipboard found.'
   finish
 endif
 
+let s:owning = {'*': -1, '+': -1}
+let s:data = {'*': [], '+': []}
+
+let g:debug_owning = s:owning
+
+let s:selection = { 'owning': -1, 'data': [] }
+
+function! s:selection.on_exit(jobid, data, event)
+  " this must be checked, as changes are nvim took ownership from itself
+  if self.owning == a:jobid
+    let self.owning = -1
+  endif
+endfunction
+
+let s:selections = { '*': s:selection, '+': copy(s:selection)} 
+
+" DEBUG:
+let g:dbg_sel = s:selections
+
 let s:clipboard = {}
 
 function! s:clipboard.get(reg)
+  if s:selections[a:reg].owning >= 0
+    echo "DEBUG: cached"
```

Comment:
Sorry forgot to mention. I have this so that it is possible to check that caching is working. It (and the other `DEBUG`) will be removed before merging.